File: api_winux.js
#!/usr/bin/env node
/**
 * API Winux - Backend Node.js/Express compatible avec nginx
 * API simple pour gérer les sessions Winux temporaires
 */

const express = require('express');
const cors = require('cors');
const fs = require('fs').promises;
const path = require('path');
const { v4: uuidv4 } = require('uuid');

const app = express();

// Configuration
const SESSIONS_FILE = '/tmp/winux_sessions.json';
const SESSION_DURATION_MINUTES = 30;
const CONTAINER_ID_MIN = 30001;
const CONTAINER_ID_MAX = 30999;

// Middleware
app.use(cors()); // Permet les requêtes cross-origin
app.use(express.json()); // Parse les requêtes JSON

/**
 * Charge les sessions depuis le fichier JSON
 */
async function loadSessions() {
    try {
        const data = await fs.readFile(SESSIONS_FILE, 'utf8');
        return JSON.parse(data);
    } catch (error) {
        if (error.code === 'ENOENT') {
            return {};
        }
        console.error('Erreur lors du chargement des sessions:', error);
        return {};
    }
}

/**
 * Sauvegarde les sessions dans le fichier JSON
 */
async function saveSessions(sessions) {
    try {
        await fs.writeFile(SESSIONS_FILE, JSON.stringify(sessions, null, 2), 'utf8');
    } catch (error) {
        console.error('Erreur lors de la sauvegarde des sessions:', error);
    }
}

/**
 * Trouve un container ID disponible
 */
function getAvailableContainerId(sessions) {
    const usedIds = new Set();
    
    for (const session of Object.values(sessions)) {
        const containerId = parseInt(session.container_id, 10);
        if (containerId >= CONTAINER_ID_MIN && containerId <= CONTAINER_ID_MAX) {
            usedIds.add(containerId);
        }
    }
    
    for (let containerId = CONTAINER_ID_MIN; containerId <= CONTAINER_ID_MAX; containerId++) {
        if (!usedIds.has(containerId)) {
            return containerId;
        }
    }
    return null;
}

/**
 * Calcule l'IP du container basée sur son ID
 */
function calculateIp(containerId) {
    // IP = 10.0.0.{container_id % 254 + 1}
    const ipLastOctet = (containerId % 254) + 1;
    return `10.0.0.${ipLastOctet}`;
}

/**
 * Calcule le nombre de secondes restantes avant expiration
 */
function calculateRemainingSeconds(expiresAtStr) {
    try {
        const expiresAt = new Date(expiresAtStr);
        const now = new Date();
        const remaining = Math.floor((expiresAt - now) / 1000);
        return Math.max(0, remaining);
    } catch (error) {
        return 0;
    }
}

/**
 * Supprime les sessions expirées
 */
async function cleanupExpiredSessions() {
    try {
        const sessions = await loadSessions();
        const now = new Date();
        const expired = [];
        
        for (const [sessionId, session] of Object.entries(sessions)) {
            try {
                const expiresAt = new Date(session.expires_at);
                if (now >= expiresAt) {
                    expired.push(sessionId);
                }
            } catch (error) {
                expired.push(sessionId);
            }
        }
        
        for (const sessionId of expired) {
            delete sessions[sessionId];
        }
        
        if (expired.length > 0) {
            await saveSessions(sessions);
            console.log(`Nettoyage: ${expired.length} session(s) expirée(s) supprimée(s)`);
        }
        
        return expired.length;
    } catch (error) {
        console.error('Erreur lors du nettoyage:', error);
        return 0;
    }
}

// Routes API

/**
 * GET /api/winux/health
 * Endpoint de santé de l'API
 */
app.get('/api/winux/health', (req, res) => {
    res.json({
        success: true,
        status: 'healthy',
        message: 'API Winux opérationnelle'
    });
});

/**
 * GET /api/winux/sessions
 * Liste toutes les sessions actives
 */
app.get('/api/winux/sessions', async (req, res) => {
    try {
        let sessions = await loadSessions();
        await cleanupExpiredSessions();
        sessions = await loadSessions(); // Recharger après nettoyage
        
        // Convertir en liste et calculer remaining_seconds
        const sessionsList = [];
        for (const [sessionId, session] of Object.entries(sessions)) {
            const remaining = calculateRemainingSeconds(session.expires_at);
            if (remaining > 0) {
                sessionsList.push({
                    ...session,
                    remaining_seconds: remaining
                });
            }
        }
        
        res.json({
            success: true,
            sessions: sessionsList
        });
    } catch (error) {
        console.error('Erreur lors de la liste des sessions:', error);
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

/**
 * POST /api/winux/sessions
 * Crée une nouvelle session Winux
 */
app.post('/api/winux/sessions', async (req, res) => {
    try {
        const durationMinutes = req.body.duration_minutes || SESSION_DURATION_MINUTES;
        
        // Vérifier qu'il n'y a pas déjà une session active
        let sessions = await loadSessions();
        await cleanupExpiredSessions();
        sessions = await loadSessions();
        
        const activeSessions = Object.values(sessions).filter(session => {
            const remaining = calculateRemainingSeconds(session.expires_at);
            return remaining > 0;
        });
        
        if (activeSessions.length > 0) {
            return res.status(400).json({
                success: false,
                error: 'Une session est déjà active. Veuillez la terminer avant d\'en créer une nouvelle.'
            });
        }
        
        // Trouver un container ID disponible
        const containerId = getAvailableContainerId(sessions);
        if (containerId === null) {
            return res.status(503).json({
                success: false,
                error: 'Aucun container ID disponible. Tous les containers sont utilisés.'
            });
        }
        
        // Créer la session
        const sessionId = uuidv4();
        const now = new Date();
        const expiresAt = new Date(now.getTime() + durationMinutes * 60 * 1000);
        
        const session = {
            session_id: sessionId,
            container_id: containerId,
            container_ip: calculateIp(containerId),
            created_at: now.toISOString(),
            expires_at: expiresAt.toISOString(),
            duration_minutes: durationMinutes,
            remaining_seconds: durationMinutes * 60
        };
        
        sessions[sessionId] = session;
        await saveSessions(sessions);
        
        console.log(`Session créée: ${sessionId} (container ${containerId}, IP ${session.container_ip})`);
        
        res.status(201).json({
            success: true,
            session: session
        });
    } catch (error) {
        console.error('Erreur lors de la création de la session:', error);
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

/**
 * GET /api/winux/sessions/:sessionId
 * Obtient les informations d'une session spécifique
 */
app.get('/api/winux/sessions/:sessionId', async (req, res) => {
    try {
        const { sessionId } = req.params;
        let sessions = await loadSessions();
        await cleanupExpiredSessions();
        sessions = await loadSessions();
        
        if (!sessions[sessionId]) {
            return res.status(404).json({
                success: false,
                error: 'Session non trouvée'
            });
        }
        
        const session = sessions[sessionId];
        const remaining = calculateRemainingSeconds(session.expires_at);
        
        if (remaining <= 0) {
            return res.status(410).json({
                success: false,
                error: 'Session expirée'
            });
        }
        
        res.json({
            success: true,
            session: {
                ...session,
                remaining_seconds: remaining
            }
        });
    } catch (error) {
        console.error('Erreur lors de la récupération de la session:', error);
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

/**
 * DELETE /api/winux/sessions/:sessionId
 * Supprime une session
 */
app.delete('/api/winux/sessions/:sessionId', async (req, res) => {
    try {
        const { sessionId } = req.params;
        const sessions = await loadSessions();
        
        if (!sessions[sessionId]) {
            return res.status(404).json({
                success: false,
                error: 'Session non trouvée'
            });
        }
        
        const session = sessions[sessionId];
        delete sessions[sessionId];
        await saveSessions(sessions);
        
        console.log(`Session supprimée: ${sessionId} (container ${session.container_id})`);
        
        res.json({
            success: true,
            message: 'Session supprimée avec succès'
        });
    } catch (error) {
        console.error('Erreur lors de la suppression de la session:', error);
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

/**
 * POST /api/winux/cleanup
 * Nettoie manuellement les sessions expirées
 */
app.post('/api/winux/cleanup', async (req, res) => {
    try {
        const count = await cleanupExpiredSessions();
        res.json({
            success: true,
            message: `${count} session(s) expirée(s) supprimée(s)`
        });
    } catch (error) {
        console.error('Erreur lors du nettoyage:', error);
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

// Nettoyage automatique toutes les minutes
setInterval(async () => {
    try {
        await cleanupExpiredSessions();
    } catch (error) {
        console.error('Erreur dans le nettoyage automatique:', error);
    }
}, 60000); // 60 secondes

// Démarrage du serveur
const PORT = process.env.PORT || 5000;
// Dans un container Docker, écouter sur 0.0.0.0 pour être accessible depuis l'extérieur
// En local, utiliser 127.0.0.1 pour la sécurité
const HOST = process.env.HOST || (process.env.NODE_ENV === 'production' ? '0.0.0.0' : '127.0.0.1');

app.listen(PORT, HOST, () => {
    console.log(`🚀 API Winux démarrée sur http://${HOST}:${PORT}`);
    console.log(`📁 Fichier de sessions: ${SESSIONS_FILE}`);
    console.log(`🌍 Mode: ${process.env.NODE_ENV || 'development'}`);
});

// Gestion des erreurs non capturées
process.on('unhandledRejection', (reason, promise) => {
    console.error('Unhandled Rejection at:', promise, 'reason:', reason);
});

process.on('uncaughtException', (error) => {
    console.error('Uncaught Exception:', error);
    process.exit(1);
});

File: api_winux.py
#!/usr/bin/env python3
"""
API Winux - Backend compatible avec nginx
API simple pour gérer les sessions Winux temporaires
"""

from flask import Flask, request, jsonify
from flask_cors import CORS
from datetime import datetime, timedelta
import json
import os
import uuid
import threading
import time

app = Flask(__name__)
CORS(app)  # Permet les requêtes cross-origin

# Configuration
SESSIONS_FILE = '/tmp/winux_sessions.json'
SESSION_DURATION_MINUTES = 30
CONTAINER_ID_MIN = 30001
CONTAINER_ID_MAX = 30999

# Structure de session
# {
#     "session_id": "uuid",
#     "container_id": 30001,
#     "container_ip": "10.0.0.1",
#     "created_at": "2025-12-05T10:00:00",
#     "expires_at": "2025-12-05T10:30:00",
#     "duration_minutes": 30,
#     "remaining_seconds": 1800
# }


def load_sessions():
    """Charge les sessions depuis le fichier JSON"""
    if os.path.exists(SESSIONS_FILE):
        try:
            with open(SESSIONS_FILE, 'r') as f:
                return json.load(f)
        except (json.JSONDecodeError, IOError):
            return {}
    return {}


def save_sessions(sessions):
    """Sauvegarde les sessions dans le fichier JSON"""
    try:
        with open(SESSIONS_FILE, 'w') as f:
            json.dump(sessions, f, indent=2)
    except IOError as e:
        print(f"Erreur lors de la sauvegarde des sessions: {e}")


def get_available_container_id(sessions):
    """Trouve un container ID disponible"""
    used_ids = {int(s.get('container_id', 0)) for s in sessions.values() 
                if s.get('container_id', 0) >= CONTAINER_ID_MIN}
    
    for container_id in range(CONTAINER_ID_MIN, CONTAINER_ID_MAX + 1):
        if container_id not in used_ids:
            return container_id
    return None


def calculate_ip(container_id):
    """Calcule l'IP du container basée sur son ID"""
    # IP = 10.0.0.{container_id % 254 + 1}
    ip_last_octet = (container_id % 254) + 1
    return f"10.0.0.{ip_last_octet}"


def calculate_remaining_seconds(expires_at_str):
    """Calcule le nombre de secondes restantes avant expiration"""
    try:
        expires_at = datetime.fromisoformat(expires_at_str)
        now = datetime.now()
        remaining = (expires_at - now).total_seconds()
        return max(0, int(remaining))
    except (ValueError, TypeError):
        return 0


def cleanup_expired_sessions():
    """Supprime les sessions expirées"""
    sessions = load_sessions()
    now = datetime.now()
    expired = []
    
    for session_id, session in sessions.items():
        try:
            expires_at = datetime.fromisoformat(session.get('expires_at', ''))
            if now >= expires_at:
                expired.append(session_id)
        except (ValueError, TypeError):
            expired.append(session_id)
    
    for session_id in expired:
        del sessions[session_id]
    
    if expired:
        save_sessions(sessions)
        print(f"Nettoyage: {len(expired)} session(s) expirée(s) supprimée(s)")
    
    return len(expired)


def cleanup_thread():
    """Thread de nettoyage automatique des sessions expirées"""
    while True:
        try:
            cleanup_expired_sessions()
        except Exception as e:
            print(f"Erreur dans le thread de nettoyage: {e}")
        time.sleep(60)  # Vérifier toutes les minutes


@app.route('/api/winux/health', methods=['GET'])
def health():
    """Endpoint de santé de l'API"""
    return jsonify({
        'success': True,
        'status': 'healthy',
        'message': 'API Winux opérationnelle'
    }), 200


@app.route('/api/winux/sessions', methods=['GET'])
def list_sessions():
    """Liste toutes les sessions actives"""
    try:
        sessions = load_sessions()
        cleanup_expired_sessions()
        sessions = load_sessions()  # Recharger après nettoyage
        
        # Convertir en liste et calculer remaining_seconds
        sessions_list = []
        for session_id, session in sessions.items():
            remaining = calculate_remaining_seconds(session.get('expires_at', ''))
            if remaining > 0:
                session_copy = session.copy()
                session_copy['remaining_seconds'] = remaining
                sessions_list.append(session_copy)
        
        return jsonify({
            'success': True,
            'sessions': sessions_list
        }), 200
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@app.route('/api/winux/sessions', methods=['POST'])
def create_session():
    """Crée une nouvelle session Winux"""
    try:
        data = request.get_json() or {}
        duration_minutes = data.get('duration_minutes', SESSION_DURATION_MINUTES)
        
        # Vérifier qu'il n'y a pas déjà une session active
        sessions = load_sessions()
        cleanup_expired_sessions()
        sessions = load_sessions()
        
        active_sessions = [s for s in sessions.values() 
                          if calculate_remaining_seconds(s.get('expires_at', '')) > 0]
        
        if active_sessions:
            return jsonify({
                'success': False,
                'error': 'Une session est déjà active. Veuillez la terminer avant d\'en créer une nouvelle.'
            }), 400
        
        # Trouver un container ID disponible
        container_id = get_available_container_id(sessions)
        if container_id is None:
            return jsonify({
                'success': False,
                'error': 'Aucun container ID disponible. Tous les containers sont utilisés.'
            }), 503
        
        # Créer la session
        session_id = str(uuid.uuid4())
        now = datetime.now()
        expires_at = now + timedelta(minutes=duration_minutes)
        
        session = {
            'session_id': session_id,
            'container_id': container_id,
            'container_ip': calculate_ip(container_id),
            'created_at': now.isoformat(),
            'expires_at': expires_at.isoformat(),
            'duration_minutes': duration_minutes,
            'remaining_seconds': duration_minutes * 60
        }
        
        sessions[session_id] = session
        save_sessions(sessions)
        
        print(f"Session créée: {session_id} (container {container_id}, IP {session['container_ip']})")
        
        return jsonify({
            'success': True,
            'session': session
        }), 201
        
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@app.route('/api/winux/sessions/<session_id>', methods=['GET'])
def get_session(session_id):
    """Obtient les informations d'une session spécifique"""
    try:
        sessions = load_sessions()
        cleanup_expired_sessions()
        sessions = load_sessions()
        
        if session_id not in sessions:
            return jsonify({
                'success': False,
                'error': 'Session non trouvée'
            }), 404
        
        session = sessions[session_id]
        remaining = calculate_remaining_seconds(session.get('expires_at', ''))
        
        if remaining <= 0:
            return jsonify({
                'success': False,
                'error': 'Session expirée'
            }), 410
        
        session_copy = session.copy()
        session_copy['remaining_seconds'] = remaining
        
        return jsonify({
            'success': True,
            'session': session_copy
        }), 200
        
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@app.route('/api/winux/sessions/<session_id>', methods=['DELETE'])
def delete_session(session_id):
    """Supprime une session"""
    try:
        sessions = load_sessions()
        
        if session_id not in sessions:
            return jsonify({
                'success': False,
                'error': 'Session non trouvée'
            }), 404
        
        session = sessions[session_id]
        del sessions[session_id]
        save_sessions(sessions)
        
        print(f"Session supprimée: {session_id} (container {session.get('container_id')})")
        
        return jsonify({
            'success': True,
            'message': 'Session supprimée avec succès'
        }), 200
        
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@app.route('/api/winux/cleanup', methods=['POST'])
def cleanup():
    """Nettoie manuellement les sessions expirées"""
    try:
        count = cleanup_expired_sessions()
        return jsonify({
            'success': True,
            'message': f'{count} session(s) expirée(s) supprimée(s)'
        }), 200
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


if __name__ == '__main__':
    # Démarrer le thread de nettoyage
    cleanup_thread_obj = threading.Thread(target=cleanup_thread, daemon=True)
    cleanup_thread_obj.start()
    
    # Démarrer l'API Flask
    # Écouter sur toutes les interfaces pour permettre le proxy nginx
    app.run(host='127.0.0.1', port=5000, debug=False)

